scheduler/core/components/optimizer/timeline.py

```diff
@@ -25,11 +26,109 @@ class Timeline:
+        return any(obs.id in self.time_slots)
+
+    def _empty_slots(self) -> np.ndarray:
+        """
+        Determine the number of empty time slots..
+        """
+        return np.where(self.time_slots == EMPTY)[0][:]
+
+    def _available_intervals(self, empty_slots: np.ndarray) -> np.ndarray:
+        """
+        Calculate the available intervals in the schedule by creating an array that contains all
+        the groups of consecutive numbers
+        """
+        return np.split(empty_slots, np.where(np.diff(empty_slots) != 1)[0]+1)
+
+    def get_available_intervals(self, first) -> np.ndarray:
+        """
+        Get the set of time_slot intervals that can be scheduled. If desired, return only the first
+        """
+        empty_slots = self._empty_slots()
+        intervals = self._available_intervals(empty_slots)
+        # print(f"length intervals = {len(intervals)}")
+        return intervals[0] if first and len(intervals) > 1 else intervals
+
+    def get_earliest_available_interval(self) -> np.ndarray:
+        """
+        Get the earliest available space in the schedule that can allocate an observation
+        """
+        return self.get_available_intervals()[0]
+
+    def add(self, iobs: int, time_slots: int, interval: int) -> datetime:
+        """
+        Add an observation index to the first open position (-1) in the given interval
+        Returns the time of this position
+        """
+        # interval = self.get_earliest_available_interval()
+        # Get first non-zero slot in given interval
+        i_first = np.where(self.time_slots[interval] == EMPTY)[0][0]
+
+        # Set values of time_slots to the observation index
+        self.time_slots[interval[i_first:i_first + time_slots]] = iobs
+
+        # Clock time for the starting index
+        start = self.start + interval[i_first] * self.time_slot_length
+
+        return start
+
+    def get_observation_order(self) -> List[Tuple[int]]:
+        """
+        Get the observation idx and position for all the schedule observations in order
+        Return
+        -------
+
+        orders: List of a 3-dimensional tuple: observation.idx, initial position
+                and last position in the plan.
+        """
+
+        schedule = self.time_slots
+        obs_comparator = self.time_slots[0]
+        start = 0
+
+        order = []
+        for position, obs_idx in enumerate(schedule):
+            if obs_idx != obs_comparator:
+                order.append((obs_comparator, start, position - 1))
+                start = position
+                obs_comparator = obs_idx
+            elif position == len(schedule) - 1:
+                order.append((obs_comparator, start, position))
+
+        return order
+
+    def __str__(self):
+        """Print the result of get_observation_order, an experiment in the use of __str__"""
+
+        obs_order = self.get_observation_order()
+
+        return f"{obs_order}"
+
+    def print(self, obsids) -> NoReturn:
```

Comment:
No type for `obsids`... should be `Sequence[ObservationID]`.
PyCharm also gets finicky with typos, so I recommend (not a requirement) to avoid warnings, using `obs_ids` and `obs_id` instead of `obsids` and `obs_id`.

Reply:
ok